Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -36,22 +36,48 @@ public struct AbstractInterpreter {
     
     public mutating func reset() {
         ifStack.removeAll()
-        stack = [VariableMap<Type>()]
+        stateDiffsStack = [VariableMap<Type>()]
         propertyTypes.removeAll()
         methodSignatures.removeAll()
     }
 
-    /// Get variables with new type which have different type in old state
-    /// Used after block end when some state on stack is removed
-    private func getStateChanges(oldState: VariableMap<Type>, newState: VariableMap<Type>) -> [(Variable, Type)] {
-        var typeChanges = [(Variable, Type)]()
-        for (variable, type) in newState {
-            let oldType = oldState[variable]
-            if oldType == nil || type != oldType  {
-                typeChanges.append((variable, type))
+    /// Update current variable types and type diffs
+    /// Used after block end when some state is removed
+    private mutating func updateStateDiffs(
+        lastStateDiff: VariableMap<Type>, typeChanges: inout [(Variable, Type)], mergeTypes: Bool = true
+    ) {
+        for (variable, oldType) in lastStateDiff {
+            var newType = oldType
+
+            if mergeTypes {
+                // If variable did not exist in parent state we keep its current type
+                // So we won't record any new type changes
+                if oldType == .nothing {
+                    newType = currentState[variable]!
+                } else {
+                    newType |= currentState[variable]!
+                }
+            }
+
+            if currentState[variable] != newType, newType != .nothing {
+                currentState[variable] = newType
```

Comment:
If `oldType == .nothing`, couldn't we remove the variable from the `currentState`?

Reply:
not really, e.g.
```
BeginIf v6
    v28 <- BinaryOperation v2, '/', v9
    // v28 = .integer | .float | .bigint
BeginElse
EndIf
// v28 = .integer | .float | .bigint
```
It would result in additional typechange.

Reply:
So in that example, we'd only want to have a single entry (just after the `BinaryOperation `) for `v28` in `Program.types` after interpretation is finished, correct?
So is the problem with `If-Else` that at the time you merge the two branches, you don't know which variables where defined inside the blocks and which ones are visible outside (because you don't merge with the "parent" state but only with a "sibling" state)? Maybe you need to do something like:
```
case BeginElse:
    let lastState = stateStack.last!
    ifStates.append(mergeCurrentState(with: lastState))
    reset(to: lastState)
    pushNewState()
case EndIf:
    ifState = popLastState(from: ifStack)
    lastState = popLastState(from: stateStack)
    mergeCurrentState(with: lastState)    // Merge Else state with state before IfElse
    mergeCurrentState(with: ifState)        // Merge resulting state with state after If to get current state
```
I.e. merge both the If and the Else state with the previous state first (which will remove all variables that are only defined in those blocks), then merge the If and the Else state together to determine the current state after the IfElse. Would something like that work?

Reply:
I don't see the difference between these 2 lines
```
lastState = popLastState(from: stateStack)
mergeCurrentState(with: lastState)    // Merge Else state with state before IfElse
```
and directly adding else branch to current state

Reply:
the point is that if you really remove variable then you have to really add it back

Reply:
Hm but why do you have to add it back? Once a block ends, if all local variables are removed from `currentState` and from the state stack (when the top entry is popped which has them all set to .nothing), then there should be no more type entries for that variable and that should be correct, shouldn't it? I think it'd be a bug to query the type of a variable after it has gone out of scope, so supporting that shouldn't be necessary. What am I missing?